content/cv/conv_and_padding.md

```diff
@@ -138,31 +148,91 @@ Bài viết này cung cấp cho bạn kiến thức cơ bản về *Convolution*
+* **Bước 1:** Khai báo thư viện, load ảnh và tạo ma trận bộ lọc. Ở đây mình dùng module [cv2.filter2D](https://opencv-python-tutroals.readthedocs.io/en/latest/py_tutorials/py_imgproc/py_filtering/py_filtering.html) cho phép *cross-correlation* và [scipy.ndimage.convolve](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.convolve.html) cho phép *convolution*.
+
+  ``` python
+  import cv2
+  import numpy as np
+  from scipy import ndimage
+  
+  # Load ảnh
+  img = cv2.imread('oencv.png',0)
+  
+  # Khởi tạo ma trận bộ lọc
+  kernel = np.asanyarray([-1,0,1,-2,0,2,-1,0,1]).reshape((3,3))
+  ```
+
+* **Bước 2:** Thực hiện *cross-correlation*
+
+  ``` python
+  corr = cv2.filter2D(img,-1, mat)
+  ```
+
+* **Bước 3:** Thực hiện *convovlution*
+
+  ``` python
+  # Chuyển ma trận ảnh về kiểu dữ liệu float
+  imgFloat = img.astype(float)
+  # Thực hiện convolution
+  conv = ndimage.convolve(imgFoat,kernel)
+  # Với output pixel < 0, đưa về giá trị 0
+  conv = np.where(conv<0, 0, kernel)
+  # Với output pixel > 255, đưa về giá trị 255
+  conv = np.where(conv>255, 255, kernel)
+  # Đưa ma trận về kiểu dữ liệu ban đầu 
+  conv = conv.astype(np.uint8)
+  ```
+
+* Kết quả:
+
+  <p>
+      <img src='https://raw.githubusercontent.com/khoaxxx238/DOC/master/img/res3.5.png' width=600>
+  </p>
+
+* Full code các bạn có thể xem ở [đây](https://github.com/khoaxxx238/DOC/blob/master/code/conv.py)
+
+## 3. Padding
 
 * Khi bạn dùng bộ lọc $W\times W$ cho một ảnh có kích thước $N\times N$, ảnh đầu ra sẽ có kích thước $(N-W+1)\times (N-W+1)$.
 
   <p>
       <img src='https://raw.githubusercontent.com/khoaxxx238/DOC/master/img/git.gif' width=500>
-      <center><caption><i>Tittle</i></caption></center>
+      <center><caption><i>Nguồn: Glass Box</i></caption></center>
   </p>
 
 * Nếu bạn muốn đầu ra là ảnh có cùng kích thước với ảnh ban đầu, bạn chỉ cần tạo vùng đệm (**pad**) xung quanh ảnh đó. Dưới đây là các kỹ thuật **padding** phổ biến:
 
-#### *3.1. Constant padding (đệm hằng số):*
+### *3.1. Constant padding (đệm hằng số):*
 
 * Giá trị *identify* của vùng đệm mang một giá trị $a$ duy nhất. Giá trị này phụ thuộc vào loại ảnh mà bạn muốn xử lý.
 
   <p>
       <img src='https://raw.githubusercontent.com/leduckhai/My-Data-Container/master/VietCS%20Blog/Nh%E1%BA%ADp%20m%C3%B4n%20Image%20filtering/Zero%20padding.png' width=300>
-      <center><caption><i>Tittle</i></caption></center>
+      <center><caption><i>Zero padding</i></caption></center>
   </p>
 
   
 
 * Nếu $a=0$, ta có **zero padding**. Điểm mạnh của **zero padding** là nó không tạo thêm đặc trưng (*feature*) cho ảnh. Đó là lý do **zero padding** được dùng nhiều trong các mô hình **Deep Learning**.
 
-#### *3.2. Reflect padding (Đệm phản chiếu)*
+* Ví dụ:
+
+  <p>
+      <img src='https://raw.githubusercontent.com/khoaxxx238/DOC/master/img/constpad.png' width=500>
+      <center><caption><i>Zero padding</i></caption></center>
+  </p>
+
+  
+
+### *3.2. Reflect padding (Đệm phản chiếu)*
 
```

Comment:
Anh nhớ trong khóa udacity và slide của các trường thì để bảo toàn dimension của ảnh sau phép convolve, nghĩa là padding là 1 cách giải quyết vấn đề của convolve nên em phải cho ví dụ trước và sau khi convolve (sử dụng các loại padding khác nhau). Như khóa Udacity em thấy đó.

Reply:
em nghĩ độc giả đọc vậy là hình dung ra rồi, phần padding này cũng không quan trọng lắm nên em nghĩ không cần viết quá chi tiết